### PR TITLE
2023-08-18 Executive Spell

### PR DESCRIPTION
**Pull request by SidestreamColdMelon**
# Description
This PR implements executive spell [planned for 2023-08-18](https://docs.google.com/spreadsheets/d/1w_z5WpqxzwreCcaveB2Ye1PP5B8QAHDglzyxKHG3CHw/edit#gid=634086607).

# Contribution Checklist
- [x] Code approved
- [x] Tests approved
- [x] CI Tests pass

# Checklist
- [x] Every contract variable/method declared as public/external private/internal
- [x] Consider if this PR needs the `officeHours` modifier override
- [x] Verify expiration (`30 days` unless otherwise specified)
- [ ] ~~Verify hash in the description matches [here](https://emn178.github.io/online-tools/keccak_256.html)~~
- [ ] ~~Validate all addresses used are in Goerli changelog or known~~
- [x] Notify any external teams affected by the spell so they have the opportunity to review
- [x] Deploy spell to Goerli `ETH_GAS_LIMIT="XXX" ETH_GAS_PRICE="YYY" make deploy`
- [x] Ensure contract is verified on `Goerli` etherscan
- [x] Change test to use Goerli spell address and deploy timestamp
- [ ] Cast spell on Goerli `make spell="0x-deployed-spell-address" cast-spell`
- [x] Run `make archive-spell` or `make date="YYYY-MM-DD" archive-spell` to make an archive directory and copy `DssSpell.sol`, `DssSpell.t.sol` and `DssSpell.t.base.sol`
- [ ] `squash and merge` this PR

### CODE REVIEW

**Comment by SidestreamColdMelon** [comment]
Current state: main spell content is ready, except for the parts that will be supplied by the Spark:
- "Transfer Spark Proxy Admin Controls" supplied via https://github.com/makerdao/spells-goerli/pull/223
  - Needs to resolve conflicts
  - CLA needs to be signed
  - Have to be reviewed by a different team familiar with the spark protocol
- "Trigger Spark Proxy Spell" prepared in https://github.com/marsfoundation/spark-spells/pull/13
  - Including it now breaks the tests, because it depends on the new admin structure
  - Have to be reviewed by a different team familiar with the spark protocol

File: src/DssSpell.sol [hunk 1]
```diff
@@ -67,91 +49,200 @@ contract DssSpellAction is DssAction {
...
+    uint256 internal constant FIVE_PT_FIVE_FIVE_PCT_RATE    = 1000000001712791360746325100;
+    uint256 internal constant FIVE_PT_EIGHT_PCT_RATE        = 1000000001787808646832390371;
+    uint256 internal constant SIX_PT_THREE_PCT_RATE         = 1000000001937312893803622469;
+    uint256 internal constant SEVEN_PCT_RATE                = 1000000002145441671308778766;
+
+    // ---------- Math ----------
+    uint256 internal constant THOUSAND = 10 ** 3;
+    uint256 internal constant MILLION  = 10 ** 6;
+    uint256 internal constant BILLION  = 10 ** 9;
+    uint256 internal constant WAD      = 10 ** 18;
```
Comment (0xdecr1pto) [inline src/DssSpell.sol:66]: `WAD` is not used, please remove it

**Comment by 0xdecr1pto** [comment]
Also `DAO Resolution for BlockTower Andromeda` is missing

File: src/DssSpell.sol (hunk 1 above)
Comment (SidestreamColdMelon) [reply src/DssSpell.sol]: Removed

**Comment by SidestreamColdMelon** [comment]
> Also `DAO Resolution for BlockTower Andromeda` is missing

It's there, right before the spark-related sections:
https://github.com/makerdao/spells-goerli/blob/fff06be6331894e1ddb2685efadf980aa139aa98/src/DssSpell.sol#L205-L209

**Comment by 0xdecr1pto** [comment]
> > Also `DAO Resolution for BlockTower Andromeda` is missing
> 
> It's there, right before the spark-related sections:
> 
> https://github.com/makerdao/spells-goerli/blob/fff06be6331894e1ddb2685efadf980aa139aa98/src/DssSpell.sol#L205-L209

Does it need to be a var? Or it's not needed for goerli?

**Comment by SidestreamColdMelon** [comment]
You're right! I've added the variable via 6938fb0

**Comment by 0xdecr1pto** [comment]
# Goerli Executive Spell Review Checklist
  ## Coding Stage

  * [x] Office Hours
    * [x] Off (default, spell casted after deploying, pause delay 60 sec.)
  * [x] Exec Hash
    * [x] empty (default)
    * [x] ensure `description` is `Goerli Spell`
  * [x] 30 Days Expiry
  * [x] `lib` (`git submodule update --init --recursive` to install/update)
    * [x] `dss-exec-lib`
      * [x] `DssExecLib.address` is unchanged by the spell PR (if changed, the new address must match the [Latest Release Tag](https://github.com/makerdao/dss-exec-lib/releases/latest) and be approved as a legitimate new release by GovAlpha)
      * [x] if submodule upgrades are present make sure `dss-exec-lib` is synced as well
        not present
      * [x] git submodule hash (run `git submodule status`) matches the latest release version or newer (NOTE: `dss-exec-lib` as installed locally will use GitHub code more recent than the 0.0.9 release)
         69b658f35d8618272cd139dfc18c5713caf6b96b lib/dss-exec-lib (v0.0.9-7-g69b658f)
         7529fa1e043b9195f41549852d9c5bb0714eaa27 lib/dss-test (heads/master)
    * [x] `dss-test`
      * [x] `dss-interfaces`
        * [x] git submodule hash matches [version used by `dss-test`](https://github.com/makerdao/dss-test/tree/master/lib) (Non-critical)
      * [x] `forge-std`
        * [ ] git submodule hash matches [version used by `dss-test`](https://github.com/makerdao/dss-test/tree/master/lib) (Non-critical)
          does not seem to be the case but is not critical
          0aa99eb8456693c015350c5e6c4f442ebe912f77 lib/forge-std (v1.3.0-3-g0aa99eb)
  * [x] `dss-interfaces`
    * [x] used in the current spell
      not used but is ok since the static intrfaces are defined
    * [x] cleanup previous ones
  * [x] Static Interfaces
    * [x] ensure they match `dss-interfaces` (Where there is a mismatch, use `cast interface` as the source of truth)
    * [x] check on-chain interface of deployed contract via `cast interface <contract_address>` to ensure correctness
    * [x] interface naming style should match with `Like` suffix (e.g. `VatLike`), with some [exceptions](https://github.com/makerdao/dss-exec-lib/blob/master/src/DssExecLib.sol#L24)
    * [x] ensure they only list used functions in spell code
  * [x] Rates match
    * [x] Compare against [IPFS](https://ipfs.io/ipfs/QmVp4mhhbwWGTfbh2BzwQB9eiBrQBKiqcPRZCaAxNUaar6)
    * [x] Check manually via `make rates pct=<pct>` (e.g. pct=0.75, for 0.75%)
       - [x] 3.58 -> 1000000001115362602336059074
       - [x] 3.33 -> 1000000001038735548426731741
       - [x] 4.08 -> 1000000001268063427242299977
       - [x] 5.00 -> 1000000001547125957863212448
       - [x] 5.25 -> 1000000001622535724756171269
       - [x] 5.55 -> 1000000001712791360746325100
       - [x] 5.80 -> 1000000001787808646832390371
       - [x] 6.30 -> 1000000001937312893803622469
       - [x] 7.00 -> 1000000002145441671308778766
    * [x] Variable name conforms to `X_PT_Y_Z_PCT_RATE` (e.g. `ZERO_PT_SEVEN_FIVE_PCT_RATE` for 0.75%)
    * [x] Variable visibility declared as `internal`
    * [x] State mutability declared as `constant`
  * [x] Math matches
    * [x] Internal Precision
      * [x] `RAD = 10 ** 45`
    * [x] Units
      * [x] `THOUSAND = 10 ** 3`
      * [x] `MILLION  = 10 ** 6`
      * [x] `BILLION  = 10 ** 9`
      * [x] Variable visibility declared as `internal`
      * [x] State mutability declared as `constant`
      * [x] Ensure they match with [config](https://github.com/makerdao/spells-mainnet/blob/master/src/test/config.sol)
   * [x] String Constants
      * [x] IPFS hashes or DAO resolutions
      * [x] Must be TOP LEVEL (or ds pause will reject)
      * [x] Exec action comments are above it
      * [x] Name matches purpose per EXEC DOC
      * [x] Visibility is `public`
      * [x] State mutability MUST BE `constant`
* [x] Stability Fee `jug.ilk.duty` ([setIlkStabilityFee](https://github.com/makerdao/dss-exec-lib/blob/v0.0.9/src/DssExecLib.sol#L792))
    * [x] ETH-A -> 3.58% `// Increase ETH-A SF by 0.14% from 3.44% to 3.58%`
    * [x] ETH-B -> 4.08% `// Increase ETH-B SF by 0.14% from 3.94%% to 4.08%`
    * [x] ETH-C -> 3.33% `// Increase ETH-C SF by 0.14% from 3.19% to 3.33%`
    * [x] WSTETH-A -> 5.25% `// Increase WSTETH-A SF by 1.81% from 3.44% to 5.25%`
    * [x] WSTETH-B -> 5.00% `// Increase WSTETH-B SF by 1.81% from 3.19% to 5.00%`
    * [x] RETH-A -> 5.25% `// Increase RETH-A SF by 1.81% from 3.44% to 5.25%`
    * [x] WBTC-A -> 5.80% `// Increase WBTC-A SF by 0.11% from 5.69% to 5.80%`
    * [x] WBTC-B -> 6.30% `// Increase WBTC-B SF by 0.11% from 6.19% to 6.30%`
    * [x] WBTC-C -> 5.55$ `// Increase WBTC-C SF by 0.11% from 5.44% to 5.55%`
* [x] Dai Savings Rate `pot.dsr` ([setDSR](https://github.com/makerdao/dss-exec-lib/blob/v0.0.9/src/DssExecLib.sol#L455))
    * [x] Double check that `PCT_RATE` is correct
      * [x] Check manually via `make rates pct=<pct>` (e.g. pct=0.75, for 0.75%)
        - [x] 5.00 -> 1000000001547125957863212448 
      * [x] Compare against [IPFS](https://ipfs.io/ipfs/QmVp4mhhbwWGTfbh2BzwQB9eiBrQBKiqcPRZCaAxNUaar6)
* [x] Debt Ceiling Changes
  * [x] Autoline Changes
    * [setIlkAutoLineDebtCeiling](https://github.com/makerdao/dss-exec-lib/blob/v0.0.9/src/DssExecLib.sol#L658)
      * [x] `WSTETH-A` -> `750 * MILLION` `// Increase WSTETH-A line by 250 million DAI from 500 million DAI to 750 million DAI (no change to gap or ttl)`
      * [x] `RETH-A` -> `75 * MILLION` `// Increase RETH-A line by 25 million DAI from 50 million DAI to 75 million DAI`
    * [setIlkAutoLineParameters](https://github.com/makerdao/dss-exec-lib/blob/v0.0.9/src/DssExecLib.sol#L648)
      * [x] `ilk: STETH-B`
      * [x] `line: 1 * BILLION` `// Increase WSTETH-B line by 500 million DAi from 500 million DAI to 1 billion DAI`
      * [x] `gap: 45 * MILLION` `// Increase WSTETH-B gap by 15 million DAI from 30 million DAI to 45 million DAI`
      * [x] `ttl: 12 hours` `        // Reduce WSTETH-B ttl by 14,400 seconds from 57,600 seconds to 43,200 seconds
`
* [x] Smart Burn Engine Parameter
    * [x] `VOW.bump` -> 20_000 DAI (RAD precision used) `// Increase vow.bump by 15,000 DAI from 5,000 DAI to 20,000 DAI` 
    * [x] `FLAP.hop` -> 6_308`// Increase hop by 4,731 seconds from 1,577 seconds to 6,308 seconds` 
* [x] RWA Updates (`RWA002-A`)
    * [x] Stability Fee `jug.ilk.duty` `// Increase RWA002-A Stability Fee by 3.5% from 3.5% to 7%`
    * [x] Liquidation Ratio `spotter.ilk.mat` `// Reduce Liquidation Ratio by 5% from 105% to 100%`
    * [x] Debt Ceiling (`line`) + Liquidation Oracle Price Bump (`val`)
        * [x] Increase Ilk Debt Ceiling (set DC + increase Global DC) `// Increase RWA002-A Debt Ceiling by 30 million DAI from 20 million DAI to 50 million DAI`
        * [x] `bump` `RwaLiquidationOracle` with new computed increased price (`val`)
          * [x] `val` should enable DAI to be drawn over the loan period while taking into account the configured `ink` amount, interest rate and liquidation ratio (see below)
            * [x] New `val` is calculated with `line * [(1 + duty) ** years] * mat` - rounded up - and makes sense in context of the [rate mechanism](https://github.com/makerdao/developerguides/blob/master/mcd/intro-rate-mechanism/intro-rate-mechanism.md). Minimum duration is usually in the Exec Doc of the spell with the RWAXXX ilk onboarding.
            * [x] Comment explaining `val` formula (`Debt ceiling * [ (1 + RWA stability fee ) ^ (minimum deal duration in years) ] * liquidation ratio`) is present
            * [x] Accompanying comment above `bump` line in format `// XXM * 1.XX^X * X.XX as a WAD` corresponding to the `val` calculation formula (e.g. `// 15M * 1.03^2 * 1.00 as a WAD`) is present along with the calculation formula on the line above
            * [x] IF combining `val` of multiple RWA ilks being combined, `val` calculation is done once per ilk and added to make the total, with workings provided in code comments. The existing `val` value can be retrieved by calling `read()` on `PIP_RWAXX` and converting the result into decimal.
            > We use only one ilk
        * [x] Poke `spotter` to pull in the new price
* [x] SubDAO Content
  * [x] SubDAO SubProxy spell execution
    * [x] SubDAO spell address matches Exec Sheet (comment or otherwise)
    * [x] SubDAO spell deployer is purely an EOA
      * [x] Ensure the deployer address is in `addresses_deployers.sol` as a comment
    * [x] Executed using `ProxyLike(SUBDAO_PROXY).exec(SUBDAO_SPELL, abi.encodeWithSignature("execute()"));`
    * [x] Execution is NOT delegate call
    * [x] Reviewer Note: Gas cost may be very high as SubDAO spells execute within the main `cast` execution. (Also note that low level call gas estimation is not done by our scripts)
  * [x] Maker Core (main spell) SubDAO actions (i.e. operate in Pause Proxy `DelegateCall` context)
    * [x] No SubDAO contract being interacted with is authed on a core contract like vat, etc. (Check comprehensively where the risk is high)
    * [x] SubDAO contract licensing and optimizations generally do not matter (except where they pose a security risk)
    * [x] SubDAO contracts and all libraries / dependencies are verified (Blocking if not true)
    * [ ] All SubDAO content addresses (i.e. provided contract addresses or EOAs) present in the Maker Core spell are present in the Exec Sheet and are correct. SubDAO addresses being authed or given any permissions and addresses being called must be confirmed by the SubDAO spell team.
> This is not true, only Spark spell in exec
    * [x] SubDAO actions match Exec Sheet (only where inline with main spell code) and do not affect core contracts
    * [x] External calls for SubDAO content are NOT delegate call
    * [x] Code does not have untoward behavior within the scope of Maker Core Contracts (e.g. up to the SubDAO proxy)
* [x] ChainLog (OK, without changes)
* [x] `addresses_goerli.sol` matches spell code
* [x] Ensure every spell variable is declared as public/internal
* [x] Ensure `immutable` visibility is only used when fetching addresses from the `ChainLog` via `DssExecLib.getChangelogAddress` and `constant` is used instead for static addresses
* [x] Fetch addresses as type `address` and wrap with `Like` suffix interfaces inline (when making calls) unless archive patterns permit otherwise (Such as `MKR`)
* [x] Use the [DssExecLib Core Address Helpers](https://github.com/makerdao/dss-exec-lib/blob/master/src/DssExecLib.sol#L166) where possible (e.g. `DssExecLib.vat()`)
* [x] Where addresses are fetched from the `ChainLog`, the variable name must match the value of the ChainLog key for that address (e.g. `MCD_VAT` rather than `vat`), except where the archive pattern differs from this pattern (e.g. MKR)
* [x] Spell actions match [GovAlpha Spell Content Sheet](https://docs.google.com/spreadsheets/d/1w_z5WpqxzwreCcaveB2Ye1PP5B8QAHDglzyxKHG3CHw)
* [x] Local Tests PASS
* [x] Ensure Good Coverage
* [x] Ensure every test function is declared as public if enabled or private if disabled

  ```

**Review by hexonaut — APPROVED** [review]
Good to deploy.

Checked:

 * Office Hours off (Always in Goerli)
 * IPFS Hash matches
 * New rate additions to rates.sol are correct
 * Rate constants match rates.sol
 * Address constants/immutables are all correct
 * Setting DSR is correct
 * Stability fee changes are correct
 * SBE Updates are correct
 * Non-DSR Parameter Changes are correct
 * Ignoring Curve stETH/ETH offboarding on Goerli
 * Ignoring delegate comp on Goerli
 * Ignoring Aave/Compound D3M Changes on Goerli
 * RWA002 Changes are correct
 * Spark Admin Transfer Reviewed by others [Skipping]
 * Spark Spell execution line is correct
 * Test config values have been updated correctly
 * Tests configured correctly with proper ones enabled
 * Tests passing

**Comment by 0xdecr1pto** [comment]
Good to deploy

```js
Running 2 tests for src/test/starknet.t.sol:StarknetTests
[PASS] testStarknet() (gas: 1659105)
[PASS] testStarknetSpell() (gas: 2324)
Test result: ok. 2 passed; 0 failed; finished in 51.57s

Running 20 tests for src/DssSpell.t.sol:DssSpellTest
[PASS] testAuth() (gas: 52110847)
[PASS] testAuthInSources() (gas: 8666152)
[PASS] testBytecodeMatches() (gas: 2877821)
[PASS] testCastCost() (gas: 1500021)
[PASS] testChainlogValues() (gas: 9850386)
[PASS] testChainlogVersionBump() (gas: 4914295)
[PASS] testContractSize() (gas: 8984)
[PASS] testDeployCost() (gas: 2862049)
[PASS] testFailNotScheduled() (gas: 14397)
[PASS] testFailTooEarly() (gas: 13607)
[PASS] testFailTooLate() (gas: 13606)
[PASS] testFailWrongDay() (gas: 13607)
[PASS] testGeneral() (gas: 39210906)
[PASS] testNextCastTime() (gas: 364790)
[PASS] testOnTime() (gas: 1495660)
[PASS] testPSMs() (gas: 2693741)
[PASS] testSparkAdminTransfer() (gas: 1536201)
[PASS] testSparkSpellIsExecuted() (gas: 1499197)
[PASS] testUseEta() (gas: 363541)
[PASS] test_RWA002_Update() (gas: 1875290)
Test result: ok. 20 passed; 0 failed; finished in 894.93s
```

**Comment by 0xdecr1pto** [comment]
## Deployed Stage

* [x] Deployed Spell is Verified
      `0x341281316C53a6c9b099581C9f87665FA5815090` OK
  * [x] Optimization Enabled: No
        No with 200 runs OK
  * [x] Other Settings: default evmVersion, GNU AGPLv3 license
        OK
* [x] Deployed Spell Code matches GitHub
      Diff is fine (execlib inlined to deployed spell)
  * [x] diffcheck etherscan source against spell PR (via `make diff-deployed-spell`)
        Diff OK, just interfaces and execlib
* [x] Deployed Spell Etherscan Checks
      Doing manually
  * [x] automated checks via `make check-deployed-spell`
        Doing manually
    * [x] verified
    * [x] license type matches
          AGPLv3 OK
    * [x] solc version matches
          0.8.16 OK
    * [x] optimizations are disabled
          No with 200 Runs OK
    * [x] dss-exec-lib library address used (under 'Libraries Used') matches the hardcoded local `DssExecLib.address` file
          `0x122f6c0dcd898b4a07310e92c3eae5d7ce0c8bb6` OK
      * [x] Check again that the PR did not modify the `DssExecLib.sol` file (e.g. look under the 'Files Changed' PR tab, etc.)
            OK
    * [x] `deployed_spell_created` matches deployment timestamp
          `1692106800` Matches OK
    * [x] `deployed_spell_block` matches deployment block number
          `9522321` Matches OK
  * [x] manual checks
      * [x] Ensure `make deploy-info tx=<tx>` matches [config](https://github.com/makerdao/spells-mainnet/blob/master/src/test/config.sol)
        * [x] `deployed_spell_created` timestamp
              1692106800 OK
        * [x] `deployed_spell_block` block number
              9522321 OK
      * [x] Ensure Etherscan `Libraries Used` matches DssExecLib [Latest Release](https://github.com/makerdao/dss-exec-lib/releases/latest)
       * [x] (For your tests to be accurate) git submodule hash matches [dss-exec-lib](https://github.com/makerdao/dss-exec-lib/releases/latest) latest release's tag commit and inspect diffs if doesn't match to ensure expected behaviour (Currently Non-Critical pending the next DssExecLib release, double check that exec lib used by the contract matches the latest release)
* [x] Archive matches `src`
  * [x] `make diff-archive-spell` for current date or or `date="YYYY-MM-DD" make diff-archive-spell` (date as per cast timestamp)
  * [x] ensure date corresponds to target exec date
* [ ] CI tests are passing
* [ ] Local Tests PASS

Waiting for local tests and CI

**Comment by 0xdecr1pto** [comment]
Good to cast
```js
Running 2 tests for src/test/starknet.t.sol:StarknetTests
[PASS] testStarknet() (gas: 1659105)
[PASS] testStarknetSpell() (gas: 2324)
Test result: ok. 2 passed; 0 failed; finished in 51.59s

Running 20 tests for src/DssSpell.t.sol:DssSpellTest
[PASS] testAuth() (gas: 52110847)
[PASS] testAuthInSources() (gas: 8666152)
[PASS] testBytecodeMatches() (gas: 2877821)
[PASS] testCastCost() (gas: 1500021)
[PASS] testChainlogValues() (gas: 9850386)
[PASS] testChainlogVersionBump() (gas: 4914295)
[PASS] testContractSize() (gas: 8984)
[PASS] testDeployCost() (gas: 2862049)
[PASS] testFailNotScheduled() (gas: 14397)
[PASS] testFailTooEarly() (gas: 13607)
[PASS] testFailTooLate() (gas: 13606)
[PASS] testFailWrongDay() (gas: 13607)
[PASS] testGeneral() (gas: 39213003)
[PASS] testNextCastTime() (gas: 364790)
[PASS] testOnTime() (gas: 1495660)
[PASS] testPSMs() (gas: 2693741)
[PASS] testSparkAdminTransfer() (gas: 1536201)
[PASS] testSparkSpellIsExecuted() (gas: 1499197)
[PASS] testUseEta() (gas: 363541)
[PASS] test_RWA002_Update() (gas: 1875290)
Test result: ok. 20 passed; 0 failed; finished in 900.81s
```

**Review by hexonaut — APPROVED** [review]
Good to cast / merge. Checked:

 * Tests passing
 * Code matches repo
 * Block and timestamp are good
 * DssExecLib matches repo address

**Review by 0xdecr1pto — APPROVED** [review]

**Pull request merged**